File: MetalTest/ImageLoader.m

#import "ImageLoader.h"
#import <Cocoa/Cocoa.h>
#import <stdlib.h>


@implementation ObjcImageLoader

- (nullable instancetype) initWithFileAtLocation:(nonnull NSURL *)location
{
  self = [super init];
  
  if (self)
  {
    NSImage* image = [[NSImage alloc] initWithContentsOfFile:location.path];
    
    NSRect imageRect = NSMakeRect(0, 0, image.size.width, image.size.height);
    CGImageRef imageRef = [image CGImageForProposedRect:&imageRect context:NULL hints:nil];
    
    // Create a suitable bitmap context for extracting the bits of the image
    _width = CGImageGetWidth(imageRef);
    _height = CGImageGetHeight(imageRef);
    NSUInteger bytesPerPixel = 4;
    _bytesPerRow = bytesPerPixel * _width;
    NSUInteger bitsPerComponent = 8;
    
    CGColorSpaceRef colorSpace = CGColorSpaceCreateDeviceRGB();
    NSMutableData *data = [NSMutableData dataWithLength:_height * _width * bytesPerPixel];
    CGContextRef context = CGBitmapContextCreate([data mutableBytes], _width, _height,
                                                 bitsPerComponent, _bytesPerRow, colorSpace,
                                                 kCGImageAlphaPremultipliedLast | kCGBitmapByteOrder32Big);
    CGColorSpaceRelease(colorSpace);
    
    // Flip the context so the positive Y axis points down
//    CGContextTranslateCTM(context, 0, _height);
//    CGContextScaleCTM(context, 1, -1);
    
    CGContextDrawImage(context, CGRectMake(0, 0, _width, _height), imageRef);
    CGContextRelease(context);
    
    _data = data;
  }
  
  return self;
}

void displayRawBuffer(const unsigned char* const data, int width, int height)
{
  return;
  
  NSBitmapImageRep* rep =
  [[NSBitmapImageRep alloc] initWithBitmapDataPlanes:&data
                                          pixelsWide:width
                                          pixelsHigh:height
                                       bitsPerSample:8
                                     samplesPerPixel:4
                                            hasAlpha:YES
                                            isPlanar:NO
                                      colorSpaceName:NSDeviceRGBColorSpace
                                         bytesPerRow:width * 4
                                        bitsPerPixel:32];
  NSImage* img = [[NSImage alloc] initWithSize:NSMakeSize(width, height)];
  [img addRepresentation:rep];
  
//  NSLog(@"%@", img);
}

void saveImage(const unsigned char* const data, int width, int height, NSString* _Nonnull path)
{
  NSBitmapImageRep* rep =
  [[NSBitmapImageRep alloc] initWithBitmapDataPlanes:&data
                                          pixelsWide:width
                                          pixelsHigh:height
                                       bitsPerSample:8
                                     samplesPerPixel:4
                                            hasAlpha:YES
                                            isPlanar:NO
                                      colorSpaceName:NSDeviceRGBColorSpace
                                         bytesPerRow:width * 4
                                        bitsPerPixel:32];
  
  NSData *nsdata = [rep representationUsingType: NSTIFFFileType properties: nil];
  [nsdata writeToFile: path atomically: NO];
}

@end

File: MetalTest/ImageLoader.swift
//
//  ImageLoader.swift
//  MetalTest
//
//  Created by Ceylo on 27/07/2017.
//  Copyright © 2017 Yalir. All rights reserved.
//

import Cocoa

@objc
class ImageLoader : NSObject {
  let width : Int
  let height : Int
  let bytesPerRow : Int
  let data : Data
  
  init(location : URL)
  {
    let image = NSImage(byReferencing: location)
    var imageRect = NSMakeRect(0.0, 0.0, image.size.width, image.size.height)
    let imageRef = image.cgImage(forProposedRect: &imageRect,
                                 context: nil, hints: nil)!
    
    self.width = imageRef.width
    self.height = imageRef.height
    let bytesPerPixel = 4
    self.bytesPerRow = bytesPerPixel * self.width
    
    let bitsPerComponent = 8
    let colorSpace = CGColorSpaceCreateDeviceRGB()
    var data = Data(count: self.height * self.bytesPerRow)
    
    data.withUnsafeMutableBytes { (bytes : UnsafeMutablePointer<UInt8>) in
      
      let bimapInfo = CGBitmapInfo(rawValue: CGImageAlphaInfo.premultipliedLast.rawValue)
        .union(.byteOrder32Big)
      
      let context = CGContext(data: bytes,
                              width: imageRef.width, height: imageRef.height,
                              bitsPerComponent: bitsPerComponent,
                              bytesPerRow: bytesPerPixel * imageRef.width,
                              space: colorSpace,
                              bitmapInfo: bimapInfo.rawValue)!
//      context.translateBy(x: 0, y: CGFloat(imageRef.height))
//      context.scaleBy(x: 1.0, y: -1.0)
      
      context.draw(imageRef, in: CGRect(x: 0, y: 0, width: imageRef.width, height: imageRef.height))
      displayRawBuffer(bytes, Int32(imageRef.width), Int32(imageRef.height))
    

//      
//      array.withUnsafeMutableBufferPointer{ (planes : inout UnsafeMutableBufferPointer<UnsafeMutablePointer<UInt8>>) in
//        
//        let bbb = UnsafeMutablePointer<UnsafeMutablePointer<UInt8>?>?(planes)
//        let bitmap = NSBitmapImageRep(bitmapDataPlanes: bbb,
//                                      pixelsWide: imageRef.width, pixelsHigh: imageRef.height,
//                                      bitsPerSample: 8,
//                                      samplesPerPixel: 4,
//                                      hasAlpha: true,
//                                      isPlanar: false,
//                                      colorSpaceName: NSDeviceRGBColorSpace,
//                                      bytesPerRow: bytesPerPixel * imageRef.width,
//                                      bitsPerPixel: 32)!
//        let nsimage = NSImage(size: NSMakeSize(CGFloat(imageRef.width), CGFloat(imageRef.height)))
//        nsimage.addRepresentation(bitmap)
//      }
    }
    
    self.data = data
  }
}
